### PR TITLE
FMWK-79 Support native Aerospike Key types (false by default)

### PR DESCRIPTION
**Pull request by roimenashe**
This PR is still a draft, investigating whether we should support (and test) the following types:
short (primitive)
Short (primitive wrapper)
byte
char

### CODE REVIEW

File: src/test/java/org/springframework/data/aerospike/assertions/KeyAssert.java [hunk 1]
```diff
@@ -15,10 +16,31 @@ public static KeyAssert assertThat(Key key) {
     }
 
     @SuppressWarnings("UnusedReturnValue")
-    public KeyAssert consistsOf(String namespace, String setName, Object userKey) {
+    public KeyAssert consistsOf(AerospikeDataSettings aerospikeDataSettings, String namespace, String setName,
+                                Object expectedUserKey) {
         Assertions.assertThat(actual.namespace).isEqualTo(namespace);
```
Comment (reugn) [inline src/test/java/org/springframework/data/aerospike/assertions/KeyAssert.java:21]: Replace this `assertj` (shouldn't be used in production code) call with something like:
```java
if (!actual.namespace.equals(namespace)) {
    throw new IllegalArgumentException("Inconsistent namespace name");
}
```

File: src/test/java/org/springframework/data/aerospike/assertions/KeyAssert.java [hunk 2]
```diff
@@ -15,10 +16,31 @@ public static KeyAssert assertThat(Key key) {
...
         Assertions.assertThat(actual.setName).isEqualTo(setName);
```
Comment (reugn) [inline src/test/java/org/springframework/data/aerospike/assertions/KeyAssert.java:22]: Same as above

File: src/test/java/org/springframework/data/aerospike/assertions/KeyAssert.java [hunk 3]
```diff
@@ -15,10 +16,31 @@ public static KeyAssert assertThat(Key key) {
...
-        Assertions.assertThat(actual.userKey.getObject()).isEqualTo(String.valueOf(userKey));
+
+        if (aerospikeDataSettings != null && aerospikeDataSettings.isKeepOriginalKeyTypes()) {
+            Assertions.assertThat(verifyActualUserKeyType(expectedUserKey)).isTrue();
+        } else {
+            // String type is used for unsupported Aerospike key types and previously for all key types in older
+            // versions of Spring Data Aerospike
+            Assertions.assertThat(checkIfActualUserKeyTypeIsString()).isTrue();
```
Comment (reugn) [inline src/test/java/org/springframework/data/aerospike/assertions/KeyAssert.java:29]: Use
```java
/*org.springframework.util.*/Assert.isTrue(checkIfActualUserKeyTypeIsString(), "Key type is not string");
```
instead.

File: src/main/java/org/springframework/data/aerospike/convert/MappingAerospikeWriteConverter.java [hunk 4]
```diff
@@ -114,15 +120,29 @@ public Optional<Key> getNewKey(AerospikeWriteData data,
...
                 String setName;
                 if (data.getSetName() != null) {
                     setName = data.getSetName();
                 } else {
                     setName = entity.getSetName();
                 }
-                return Optional.of(new Key(data.getNamespace(), setName, id));
+
+                // Store record key as it is (if Aerospike supports it natively and configured)
+                if (aerospikeDataSettings.isKeepOriginalKeyTypes() &&
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/convert/MappingAerospikeWriteConverter.java:131]: Take the `&&` operator to the new line

File: src/main/java/org/springframework/data/aerospike/convert/MappingAerospikeWriteConverter.java [hunk 5]
```diff
@@ -114,15 +120,29 @@ public Optional<Key> getNewKey(AerospikeWriteData data,
...
             AerospikePersistentProperty idProperty = entity.getIdProperty();
             if (idProperty != null) {
-                String id = accessor.getProperty(idProperty, String.class); // currently id can only be a String
-                Assert.notNull(id, "Id must not be null!");
                 String setName;
                 if (data.getSetName() != null) {
                     setName = data.getSetName();
                 } else {
                     setName = entity.getSetName();
                 }
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/convert/MappingAerospikeWriteConverter.java:128]: nit: can be simplified with
```java
String setName = Optional.ofNullable(data.getSetName()).orElse(entity.getSetName());
```

File: src/main/java/org/springframework/data/aerospike/convert/MappingAerospikeWriteConverter.java [hunk 6]
```diff
@@ -287,4 +313,37 @@ private int getExpirationFromProperty(ConvertingPropertyAccessor<?> accessor,
...
+            return new Key(namespace, set, (long) userKey);
+        } else if (userKey.getClass() == Character.class) {
+            return new Key(namespace, set, (char) userKey);
+        } else if (userKey.getClass() == Byte.class) {
+            return new Key(namespace, set, (byte) userKey);
+        } else if (userKey.getClass() == byte[].class) {
+            return new Key(namespace, set, (byte[]) userKey);
+        }
+        // Could not construct a key of native supported Aerospike record key type
+        return null;
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/convert/MappingAerospikeWriteConverter.java:343]: Why not supporting Double/Float?

Comment (roimenashe) [reply src/main/java/org/springframework/data/aerospike/convert/MappingAerospikeWriteConverter.java:343]: Aerospike Record Key doesn't support Double/Float.

**Review by reugn — APPROVED** [review]

File: src/main/java/org/springframework/data/aerospike/config/AerospikeDataSettings.java [hunk 7]
```diff
@@ -32,6 +32,10 @@ public class AerospikeDataSettings {
     int indexCacheRefreshFrequencySeconds = 3600;
     @Builder.Default
     long queryMaxRecords = 10_000L;
+    // When true, stores key types (record keys and map keys) as original type if supported. False means always store
```
Comment (agrgr) [inline src/main/java/org/springframework/data/aerospike/config/AerospikeDataSettings.java:35]: Define how @Id fields (primary keys) and Map keys are stored: false - always as String, true - preserve original type if it is supported

Used this in the documentation, what do you think?

Comment (roimenashe) [reply src/main/java/org/springframework/data/aerospike/config/AerospikeDataSettings.java:35]: I like it

**Review by agrgr — APPROVED** [review]

**Pull request merged**